doc/man/man3/crypto_lock.3monocypher

```diff
@@ -226,18 +226,36 @@ attacker's interference.
 .Fa plain_text
 does not need to be wiped if the decryption fails.
 .Sh EXAMPLES
+The following examples assume that a function called
+.Fn random_bytes
+exists.
+It fills the given buffer with cryptographically secure random
+bytes
+(see
+.Xr intro 3monocypher
+for some advice on how to accomplish this yourself).
+The function has this prototype:
+.Ft void
+.Fo random_bytes
+.Fa "uint8_t *buf"
+.Fa "size_t len"
+.Fc
+.Pp
```

Comment:
I believe those are overkill, and best omitted. Instead, a short *code* comment in the example itself should do:

    random_bytes(key, 32) /* use the OS to implement this function */

Alternatively, we could just give up and use arc4random_buf(3), which is my recommended API whenever available anyway.

Reply:
> Alternatively, we could just give up and use arc4random_buf(3), which is my recommended API whenever available anyway.

As much as I would love to do that (and have done that for the custom hash function documentation because I was assuming the platform as a whole to have an alternative SHA-2 implementation to test with), I don't think that's a good idea for the *general-purpose* documentation. Please recall that the dominant systems are Windows, macOS and Linux, of which only macOS provides arc4random_buf(3). Ignoring the embedded non-Linux case (bare metal, custom OS, VxWorks come to mind), arc4random_buf(3) is provided by niche operating systems only ({Free,Open,Net}BSD, illumos, Solaris). So in the most common cases, there's a good chance that people will run into a snag.

Said snag is also a “hard” problem: They will be faced with a compilation error for arc4random_buf(3) being missing, rather than a more obviously named random_bytes() function. The former might make them look into all kinds of wrong things (“Am I missing something?”) rather than re-checking the comment (ideally) or googling a more generically-useful query that hopefully leads to a StackOverflow post that is sufficiently generic. I admit this is speculative, but I'd rather avoid having a *specific* problem that stands in for misunderstanding a *generic* problem.

I do think I'll roll with the comment instead of arc4random() and take out the paragraph to keep it concise.

Reply:
My Ubuntu seems to provide arc4random(3). At least there's the man page. Windows users are indeed left in the dust.

> Said snag is also a “hard” problem: They will be faced with a compilation error for arc4random_buf(3) being missing, rather than a more obviously named random_bytes() function. 

Ah, that's a problem indeed. `random_bytes()` it is, then.

> I do think I'll roll with the comment instead of arc4random() and take out the paragraph to keep it concise.

Works for me.

Reply:
> My Ubuntu seems to provide arc4random(3).

For the sake of completeness:

It provides libbsd as part of the libbsd0 package due to dependencies of other packages that were installed by default (`apt-cache rdepends --installed libbsd0`, which for your machine will probably include xserver-xorg-core, libedit2 [via openssh-client] and maybe samba). Using libbsd is a bit complicated; looking at libbsd(7), you either prefix your includes like so: `#include <bsd/unistd.h>` (which breaks on the BSDs themselves, or you use it in "overlay mode", compiling with `-isystem /usr/include/bsd -DLIBBSD_OVERLAY` (assuming the installation prefix was `/usr`, cf. `pkg-config --cflags libbsd-overlay`) and linking with `-lbsd` at the end in either case. You cannot rely on libbsd existing on all distributions, however; it'll be routinely missing from e.g. Alpine Linux.

Here's a handy table summarizing what I know at the top of my head:

Mechanism | OS | Notes
-|-|-
getrandom(2) | Linux, FreeBSD, illumos, Solaris | recent Linux invention; illumos tells you to prefer arc4random; fairly recent invention; can fail; OpenBSD resists adoption
getentropy() | OpenBSD, Linux, FreeBSD, NetBSD, illumos, Solaris | older OpenBSD invention; only supposed to be used to *seed* an RNG (used by arc4random internally); on standardization track for POSIX; emulated on Linux in glibc and is hence getentropy(*3*) there; limit of 256 bytes per invocation; can fail
arc4random(3) | every BSD, macOS, illumos, Solaris | OpenBSD invention, kernel-seeded userspace PRNG that reseeds itself periodically (which is the reason getentropy(2) was invented in the first place, since it originally used `/dev/urandom`); early versions used actual RC4 and had statistical bias accordingly (still affects FreeBSD 11 [!] but not 12); does not fail and instead aborts the program if getting entropy from kernel fails (the kernel is supposed to guarantee delivery of entropy as part of the contract with the arc4random interface); part of libc, not the kernel interface
`/dev/urandom` | almost every *NIX like including stuff like HP-UX | can fail in a myriad of ways and be missing in a chroot, but at least it's available almost everywhere
BCryptGenRandom() | Windows | The way you're supposed to do it on Windows; can actually operate without an algorithm handle apparently (unlike the XP-compatible CryptGenRandom); can fail
RtlGenRandom() | Windows | The way libsodium does it (whose comments say so do Rust, Firefox and BoringSSL)

Reply:
> My Ubuntu seems to provide arc4random(3).

For the sake of completeness:

It provides libbsd as part of the libbsd0 package due to dependencies of other packages that were installed by default (`apt-cache rdepends --installed libbsd0`, which for your machine will probably include xserver-xorg-core, libedit2 [via openssh-client] and maybe samba). Using libbsd is a bit complicated; looking at libbsd(7), you either prefix your includes like so: `#include <bsd/unistd.h>` (which breaks on the BSDs themselves, or you use it in "overlay mode", compiling with `-isystem /usr/include/bsd -DLIBBSD_OVERLAY` (assuming the installation prefix was `/usr`, cf. `pkg-config --cflags libbsd-overlay`) and linking with `-lbsd` at the end in either case. You cannot rely on libbsd existing on all distributions, however; it'll be routinely missing from e.g. Alpine Linux.

Here's a handy table summarizing what I know at the top of my head:

Mechanism | OS | Notes
-|-|-
getrandom(2) | Linux, FreeBSD, illumos, Solaris | recent Linux invention; illumos tells you to prefer arc4random; fairly recent invention; can fail; OpenBSD resists adoption
getentropy() | OpenBSD, Linux, FreeBSD, NetBSD, illumos, Solaris | older OpenBSD invention; only supposed to be used to *seed* an RNG (used by arc4random internally); on standardization track for POSIX; emulated on Linux in glibc and is hence getentropy(*3*) there; limit of 256 bytes per invocation; can fail
arc4random(3) | every BSD, macOS, illumos, Solaris | OpenBSD invention, kernel-seeded userspace PRNG that reseeds itself periodically (which is the reason getentropy(2) was invented in the first place, since it originally used `/dev/urandom`); early versions used actual RC4 and had statistical bias accordingly (still affects FreeBSD 11 [!] but not 12); does not fail and instead aborts the program if getting entropy from kernel fails (the kernel is supposed to guarantee delivery of entropy as part of the contract with the arc4random interface); part of libc, not the kernel interface
`/dev/urandom` | almost every *NIX like including stuff like HP-UX | can fail in a myriad of ways and be missing in a chroot, but at least it's available almost everywhere
BCryptGenRandom() | Windows | The way you're supposed to do it on Windows; can actually operate without an algorithm handle apparently (unlike the XP-compatible CryptGenRandom); can fail
RtlGenRandom() | Windows | The way libsodium does it (whose comments say so do Rust, Firefox and BoringSSL)